### PR TITLE
feat(storage): add Autoclass support

### LINKED ISSUE #6592: storage: add support for Autoclass
Autoclass feature at the bucket level, allows for the automatic selection of the best storage class based on object metadata and access patterns. Currently, Autoclass supports  **new buckets** only - a bucket can be enabled for Autoclass only at creation, once disabled it cannot be enabled again. See [go/gcs-dpe-autoclass](http://go/gcs-dpe-autoclass) (internal) for more details.

**System tests are only required in languages with code changes other than documentation:**
- Autoclass can be enabled via Bucket `Create`
- Autoclass can be disabled via Bucket `Patch` and `Update` operations
- Bucket `Get` operations should return the correct state of Autoclass on the bucket which includes the enabled boolean field and toggleTime timestamp field

Please wait for approval from myself before merging/releasing so we can coordinate around release dates.

### PR DESCRIPTION
**Pull request by cojenco**
Add support for Autoclass, which allows for automatic selection of the best storage class based on object access patterns

Fixes #6592

### CODE REVIEW

File: storage/bucket.go
Comment (tritone) [inline storage/bucket.go:1953]: This is a bit confusing, so maybe add a comment that we are only parsing `ToggleTime` if available. We should also check in integration tests that it is parsed successfully (ie that it is non-zero).

File: storage/bucket.go
Comment (tritone) [inline storage/bucket.go:1107]: Remember to fill this in before we merge.

File: storage/bucket.go
Comment (tritone) [inline storage/bucket.go:1928]: Can leave "bucket" out of these method names (so `toRawAutoclass` and `toProtoAutoclass`)

File: storage/bucket.go
Comment (tritone) [inline storage/bucket.go:1948]: Let's say `toAutoclassFromRaw` (mirroring customPlacement)

File: storage/bucket_test.go
Comment (tritone) [inline storage/bucket_test.go:60]: Let's make sure we have test coverage for the proto converters as well. Also, you should make an integration test to cover the roundtrip with the service (both create and update).

**Review by tritone — COMMENTED** [review]
Nice start on this Cathy!

File: storage/bucket.go
Comment (cojenco) [reply storage/bucket.go:1966]: Comment added. I also updated the logic to match how RetentionPolicy handles valid/non-zero timestamps, per https://github.com/googleapis/google-cloud-go/pull/6807

File: storage/bucket.go
Comment (cojenco) [reply storage/bucket.go]: done ;)

File: storage/bucket.go
Comment (cojenco) [reply storage/bucket.go]: done

File: storage/bucket_test.go
Comment (cojenco) [reply storage/bucket_test.go:62]: Yep was meaning to write an integration test... also added 2 tests for the proto converters. @tritone PTAL, thanks!

Comment (tritone) [reply storage/bucket_test.go:62]: Thanks for adding these additional tests; didn't realize we were missing these unit tests for proto converter methods. We should consider auditing and filling others in if they are missing. @noahdietz @BrennaEpp

File: storage/bucket.go
Comment (tritone) [reply storage/bucket.go:1966]: Hmm, I actually think we should avoid returning an error if `ToggleTime` is not present. In general we want to avoid returning errors when the response from the server doesn't look exactly like we would expect; IMO it is to ignore the field rather than have this client-side validation on the response, in case the API is later updated in a way we don't expect.

File: storage/bucket_test.go
Comment (tritone) [inline storage/bucket_test.go:715]: Thanks very much for adding this! We should have done it sooner.

File: storage/integration_test.go
Comment (cojenco) [reply storage/integration_test.go]: Done, thanks!

File: storage/bucket.go
Comment (cojenco) [reply storage/bucket.go:1962]: Agreed that we want to avoid client-side validations on the response. I've updated the converter so that the field is ignored in those cases.

We'd want to revisit how we're handling the read only timestamp for RetentionPolicy as well. Looking at our code base, that requires a bigger scope of change, so I'm separating that into an independent tracking issue.

**Review by tritone — APPROVED** [review]
One test comment, otherwise LGTM pending feature release.

**Pull request merged**
